### PR TITLE
STM32WB0: add DMA support

### PR DESCRIPTION
**Pull request by mathieuchopstm**
Add support for DMA+DMAMUX on STM32WB0 series hardware and updates the boards' YAML accordingly; also add a test overlay for Nucleo-WB09KE in relevant DMA tests so that correct behavior can be verified.

### CODE REVIEW

File: drivers/dma/dmamux_stm32.c [hunk 1]
```diff
@@ -76,6 +76,14 @@ uint32_t table_ll_channel[] = {
 #else
 #define dmamux_channel_typedef const DMAMUX_Channel_TypeDef
 #endif
+
+/**
+ * STM32WB0 series DMAMUX lacks synchronization inputs and request generator.
+ * For this reason, the following functions are not present in STM32CubeWB0 LL.
+ *
+ * @note As of writing, these functions are not used by the STM32 driver.
```
Comment (erwango) [inline drivers/dma/dmamux_stm32.c:84]: If not used, then let's clean them up.

File: drivers/dma/dma_stm32.c [hunk 2]
```diff
@@ -780,6 +780,7 @@ static void dma_stm32_config_irq_0(const struct device *dev)
 	ARG_UNUSED(dev);
 
 	DMA_STM32_IRQ_CONNECT(0, 0);
+#if !defined(CONFIG_SOC_SERIES_STM32WB0X) /* Only 1 IRQ on STM32WB0 */
```
Comment (erwango) [inline drivers/dma/dma_stm32.c:783]: Can't use `#if DT_INST_IRQ_HAS_IDX(0, 1)` ?

File: drivers/dma/dma_stm32.c [hunk 3]
```diff
@@ -30,7 +30,7 @@ LOG_MODULE_REGISTER(dma_stm32, CONFIG_DMA_LOG_LEVEL);
 #endif
 
 #if DT_NODE_HAS_STATUS_OKAY(DT_DRV_INST(0))
-#if DT_INST_IRQ_HAS_IDX(0, 7)
+#if DT_INST_IRQ_HAS_IDX(0, 7) || defined(CONFIG_SOC_SERIES_STM32WB0X)
```
Comment (erwango) [inline drivers/dma/dma_stm32.c:33]: Some information (could be in commit message) about why a `CONFIG_SOC_SERIES_STM32WB0X` specific rule is needed would be welcome.

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/dma/dma_stm32.c (hunk 2 above)
Comment (mathieuchopstm) [reply drivers/dma/dma_stm32.c:783]: Indeed, it should work. Changing.

File: drivers/dma/dma_stm32.c (hunk 3 above)
Comment (mathieuchopstm) [reply drivers/dma/dma_stm32.c:39]: Added a comment above.

File: drivers/dma/dmamux_stm32.c (hunk 1 above)
Comment (mathieuchopstm) [reply drivers/dma/dmamux_stm32.c]: Done.

File: drivers/dma/dmamux_stm32.c [hunk 4]
```diff
@@ -56,41 +56,12 @@ struct dmamux_stm32_config {
...
-uint32_t table_ll_channel[] = {
-	LISTIFY(DT_INST_PROP(0, dma_channels), DMAMUX_CHANNEL, (,))
-};
-
 #if !defined(CONFIG_SOC_SERIES_STM32G0X) && \
 	!defined(CONFIG_SOC_SERIES_STM32C0X)
 #define dmamux_channel_typedef DMAMUX_Channel_TypeDef
 #else
 #define dmamux_channel_typedef const DMAMUX_Channel_TypeDef
 #endif
```
Comment (erwango) [inline drivers/dma/dmamux_stm32.c:64]: Probably this could be removed as well.

**Review by erwango — CHANGES_REQUESTED** [review]
Otherwise LGTM

**Comment by mathieuchopstm** [comment]
Patch V2:
* remove remaining unused piece of code in DMAMUX driver
* refactor DMA driver to reduce per-series crust and make intents more explicit
* use proper binding and `interrupts` property in DTS

**Comment by mathieuchopstm** [comment]
Fixed wrong `#dma-cells` value in DTS + split DMAMUX driver modification to separate commit

**Review by erwango — DISMISSED** [review]

File: dts/arm/st/wb0/stm32wb0.dtsi [hunk 5]
```diff
@@ -217,6 +217,31 @@
 			interrupts = <7 0>;
 			status = "disabled";
 		};
+
+		dma1: dma@48700000 {
+			compatible = "st,stm32-dma-v2bis";
+			#dma-cells = <2>;
+			reg = <0x48700000 256>;
+			clocks = <&rcc STM32_CLOCK_BUS_AHB0 (1 << 0)>;
```
Comment (gautierg-st) [inline dts/arm/st/wb0/stm32wb0.dtsi:225]: Now that #79683 has been merged, this node should be updated

File: dts/arm/st/wb0/stm32wb0.dtsi [hunk 6]
```diff
@@ -217,6 +217,31 @@
...
+			status = "disabled";
+		};
+
+		dmamux1: dmamux@48800000 {
+			compatible = "st,stm32-dmamux";
+			reg = <0x48800000 DT_SIZE_K(1)>;
+			/* `clocks` property is identical between DMA and DMAMUX
+			 * because they share a single common bit in RCC registers
+			 */
+			clocks = <&rcc STM32_CLOCK_BUS_AHB0 (1 << 0)>;
```
Comment (gautierg-st) [inline dts/arm/st/wb0/stm32wb0.dtsi:238]: Now that #79683 has been merged, this node should be updated

**Review by gautierg-st — CHANGES_REQUESTED** [review]

**Review by gautierg-st — APPROVED** [review]

File: dts/arm/st/wb0/stm32wb0.dtsi (hunk 5 above)
Comment (mathieuchopstm) [reply dts/arm/st/wb0/stm32wb0.dtsi]: Done + added a commit to update nodes missed by the PR (I2C/SPI)

File: dts/arm/st/wb0/stm32wb0.dtsi (hunk 6 above)
Comment (mathieuchopstm) [reply dts/arm/st/wb0/stm32wb0.dtsi]: Done

**Review by erwango — APPROVED** [review]

**Pull request merged**